### DISCUSSION TITLE
check_tasksched - How to get last Event-ID

### DISCUSSION THREAD

**Discussion by k-tf** [mickem/nscp#833] [Q&A] [answered]
Hi there!

I'm currently trying to check for stuck tasks on the _Windows Task Scheduler_ with **check_tasksched.**

The first idea was to refer to `most_recent_run_time` but that is being reset when the next schedule occurs, even if that one fails with _Launch request ignored, instance already running_ (Event-ID 322).

So my problem could be solved in either of two ways:

1. Get a correct value for `most_recent_run_time` that refers to the time when the current task was _really_ started (and not the time the scheduler did _try_ to start a task)
2. Get the value of the last Event-ID for this task. (If it's 322, another still running task is blocking the execution of a new one)

Does somebody got an idea on how to accomplish this? Or would this a _new feature request_?

**Comment by k-tf** [comment, marked as answer]
Self-answered:

after tedious testing I verified the following return codes:

- **267009**	The task is currently running
- **-2147020576** - The scheduler didn't start a new task because there's already an instance of this task running. According to the Windows documentation this was supposed to be 2147750687 but not on our Windows 2016 Server and -2147020576 is also _not_ a two-complement of 2147750687. I gave up trying to understand this.

So for my needs I trigger critical if the task 

- is running longer than 50 Minutes, 
- OR the exit code is negative (catches any weird exit code)
- OR the last schedule is longer ago than 70 minutes (60 Minutes is our default schedule and I add 10 Minutes to avoid false positives when the schedule is late because of heavy system load)

The final check command looks like this:
`
check_tasksched "filter=title eq 'MYTASK'" "critical=(exit_code = 267009 AND most_recent_run_time < -50m) OR (exit_code < 0) OR (most_recent_run_time < -70m)" "warning=exit_code = 267009 AND most_recent_run_time < -40m" "empty-state=critical" "empty-syntax=CRITICAL: Task MYTASK not found" "detail-syntax=%(title): exit_code=%(exit_code)  most_recent_run_time=%(most_recent_run_time) GMT" show-all
`

**Comment by mickem** [comment]
Thanks for coming back with the answer, and your reading of the codes is right. For anyone finding this later:

| `exit_code` | Hex | Meaning |
|---|---|---|
| `267009` | `0x41301` | `SCHED_S_TASK_RUNNING`: the task is currently running |
| `-2147020576` | `0x800710E0` | HRESULT for Win32 error 4320, `ERROR_REQUEST_REFUSED` ("The operator or administrator has refused the request"). This is what Task Scheduler stores as the last result when it refuses to launch because an instance is already running (event 322). |

`exit_code` is a signed 32-bit value, which is why every Windows error HRESULT (`0x8...`) shows up negative and why the default `critical=exit_code < 0` catches them.

On `most_recent_run_time`: it is `IRegisteredTask::LastRunTime` straight from the Task Scheduler API, and Windows updates it on every launch attempt, including the refused ones. The API has no "last event id" per task, so the check cannot offer one; the events live in the `Microsoft-Windows-TaskScheduler/Operational` channel. If you want to alert on event 322 itself, `check_eventlog` can read that channel (task history has to be enabled on the host):

```
check_eventlog "file=Microsoft-Windows-TaskScheduler/Operational" "filter=id = 322 and written > -1h" "detail-syntax=${message}"
```

Recent releases also added two keywords that make your rule a bit more readable than comparing against the raw codes:

- `task_status`: the task state as a word (`ready`, `running`, `disabled`, `queued`), so `task_status = 'running'` instead of `exit_code = 267009`.
- `last_run_age`: seconds since the task last ran (`-1` if it never ran).

So the equivalent of your check on a current release is:

```
check_tasksched "filter=title = 'MyTask'" "critical=(task_status = 'running' AND last_run_age > 3000) OR exit_code < 0 OR last_run_age > 4200"
```

Your `most_recent_run_time < -50m` form keeps working too; the relative-time comparison is the same thing expressed as a date.
